lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -569,6 +569,20 @@ func (d *zfs) CreateVolumeFromCopy(vol Volume, srcVol Volume, copySnapshots bool
 		}
 	}
 
+	// When not allowing inconsistent copies and the volume has a mounted filesystem, we must ensure it is
+	// consistent by syncing and freezing the filesystem to ensure unwritten pages are flushed and that no
+	// further modifications occur while taking the source snapshot.
+	var unfreezeFS func() error
+	sourcePath := srcVol.MountPath()
+	if !allowInconsistent && srcVol.contentType == ContentTypeFS && srcVol.IsBlockBacked() && filesystem.IsMountPoint(sourcePath) {
+		unfreezeFS, err := d.filesystemFreeze(sourcePath)
```

Comment:
This looks wrong. This `unfreezeFS` variable is only valid within the `if`-scope, and the check below (`if unfreezeFS != nil`) will always be false. You should rather do

```go
var unfreezeFS func() error
var err error

if ... {
    unfreezeFS, err = d.filesystemFreeze(sourcePath)
}

Reply:
Good spot!

Reply:
@monstermunchkin fixed:

```
DEBUG  [2023-03-02T11:46:10Z] CreateInstanceFromCopy started                instance=c2 project=default snapshots=false src=c1
DEBUG  [2023-03-02T11:46:10Z] CreateInstanceFromCopy same-pool mode detected  instance=c2 project=default snapshots=false src=c1
INFO   [2023-03-02T11:46:10Z] Filesystem frozen                             driver=zfs path=/var/lib/lxd/storage-pools/foo1/containers/c1 pool=foo1
INFO   [2023-03-02T11:46:10Z] Filesystem unfrozen                           driver=zfs path=/var/lib/lxd/storage-pools/foo1/containers/c1 pool=foo1
DEBUG  [2023-03-02T11:46:11Z] Activated ZFS volume                          dev=foo1/containers/c2 driver=zfs pool=foo1 volName=c2
DEBUG  [2023-03-02T11:46:11Z] Regenerating filesystem UUID                  dev=/dev/zvol/foo1/containers/c2 driver=zfs fs=xfs pool=foo1
DEBUG  [2023-03-02T11:46:11Z] Mounted ZFS volume                            dev=foo1/containers/c2 driver=zfs path=/var/lib/lxd/storage-pools/foo1/containers/c2 pool=foo1 volName=c2
DEBUG  [2023-03-02T11:46:11Z] Unmounted ZFS volume                          dev=foo1/containers/c2 driver=zfs path=/var/lib/lxd/storage-pools/foo1/containers/c2 pool=foo1 volName=c2
DEBUG  [2023-03-02T11:46:11Z] Deactivated ZFS volume                        dev=foo1/containers/c2 driver=zfs pool=foo1 volName=c2
DEBUG  [2023-03-02T11:46:11Z] CreateInstanceFromCopy finished               instance=c2 project=default snapshots=false src=c1
```